src/components/ModalWithForm/ModalWithForm.js

```diff
@@ -0,0 +1,53 @@
+import React, { useEffect } from "react";
+
+const ModalWithForm = ({
+  buttonText,
+  name,
+  title,
+  children,
+  isOpen,
+  onClose,
+  onSubmit,
+  isValid,
+}) => {
+  const handleCloseModalOverlayClick = (e) => {
+    if (e.target === e.currentTarget) {
+      onClose();
+    }
+  };
+
+  useEffect(() => {
+    const handleEscKey = (e) => {
+      if (e.key === "Escape" && isOpen) {
+        onClose();
+      }
+    };
+
+    document.addEventListener("keydown", handleEscKey);
```

Comment:
You need to add an `Escape` listener like this to be able to remove it **correctly** when modals are closed:

```jsx
   useEffect(() => {

    if (!isOpen) return; // stop the effect not to add the listener if there is no active modal

    const handleEscClose = (e) => {  // define the function inside useEffect not to lose the reference on rerendering
      if (e.key === "Escape") {
        onClose();
      }
    };

    document.addEventListener("keydown", handleEscClose);

    return () => {  // don't forget to add a clean up function for removing the listener
      document.removeEventListener("keydown", handleEscClose);
    };
  }, [isOpen]);  // watch isOpenhere
```

Here you define function ` handleEscClose` inside the `useEffect` so that it could maintain a **fixed** reference if the component rerenders.
And of course we need to return a `clean up` function to **`remove`** the listener if the component unmounts.
And don’t forget about the **`dependencies array`** in order to watch `isOpen`